src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,151 @@
+import cromwell.parser.WdlParser.{Ast, AstList}
+
+import scala.annotation.tailrec
+import scala.collection.mutable.{Map, HashMap}
+import scala.collection.JavaConverters._
+
+object Scope {
+
+  /**
+   * Generate a Scope Tree from an Ast.
+   * Ast must match a Scope (currently: Workflow, Scatter or Call)
+   * @throws UnsupportedOperationException if node does not match any supported Scopes
+   * @return generated Scope Tree.
+   */
+  def generateTree(node: Ast, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Scope = {
+
+    val scopeIndexes: Map[Class[_ <: Scope], Int] = HashMap.empty
+
+    /**
+     * Retrieve the list of children Asts from the AST body.
+     * Return empty seq if body is null or is not a list.
+     */
+    def getChildrenList(ast: Ast): Seq[Ast] = Option(ast.getAttribute("body")) collect {
+      case list: AstList => list.asScala.toVector
+    } getOrElse Seq.empty collect {
+      case ast: Ast => ast
+    }
+
+    /**
+     * Generate a scope from the ast parameter and all its descendants recursively.
+     * @return Some(instance of Scope) if the node has a Scope equivalent, None otherwise
+     */
+    def generateScopeTree(node: Ast, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter, parent: Option[Scope]): Option[Scope] = {
+      val scope: Option[Scope] = node match {
+        case w: Ast if w.getName == AstNodeName.Workflow => Option(Workflow(w, wdlSyntaxErrorFormatter, parent))
+        case c: Ast if c.getName == AstNodeName.Call => Option(Call(c, namespaces, tasks, wdlSyntaxErrorFormatter, parent))
+        case s: Ast if s.getName == AstNodeName.Scatter => Option(Scatter(s, scopeIndexes.getOrElse(classOf[Scatter], 0), parent))
+
+        /* Cases need to be added here to support new scopes */
+        case _ => None
+      }
+      //Generate and set children recursively
+      scope foreach { sc =>
+        //Increment index if needed
+        scopeIndexes.put(sc.getClass, scopeIndexes.getOrElse(sc.getClass, 0) + 1)
+        sc.setChildren(getChildrenList(node) flatMap { generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter, Option(sc)) })
+      }
+      scope
+    }
+
+    generateScopeTree(node, namespaces, tasks, wdlSyntaxErrorFormatter, None)
+      .getOrElse(throw new UnsupportedOperationException("Input node cannot be instantiated as a Scope"))
+  }
+
+  /**
+   * Generate a workflow from an Ast. The ast must be one of a workflow.
+   * @throws UnsupportedOperationException if the ast is not a workflow ast
+   * @return a workflow
+   */
+  def generateWorkflow(ast: Ast, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Workflow = {
+    generateTree(ast, namespaces, tasks, wdlSyntaxErrorFormatter) match {
+      case w: Workflow => w
+      case _ => throw new UnsupportedOperationException("Ast is not a 'Workflow Ast'")
```

Comment:
Looking at it again I was thinking something like : 
```scala
ast.getName match {
      case AstNodeName.Workflow => generateTree(ast, namespaces, tasks, wdlSyntaxErrorFormatter).asInstanceOf[Workflow]
      case nonWorkflowAst => throw new UnsupportedOperationException(s"Ast is not a 'Workflow Ast' but a '$nonWorkflowAst Ast'")
    }
```
So we don't even generate the tree if the ast is not a AstNodeName.Workflow and throw an exception right away, instead of generating it anyway and check afterwards if it has the right type. I don't know if printing the whole Ast in the error message is necessary, maybe just the name ?